src/states/infiniteweightpeps.jl

```diff
@@ -0,0 +1,211 @@
+
+"""
+    const PEPSWeight{S}
+
+Default type for PEPS bond weights with 2 virtual indices,
+conventionally ordered as: ``wt : ES ← WN``. 
+Here, `ES`, `WN` denote the east/south, west/north spaces, respectively.
+"""
+const PEPSWeight{S} = AbstractTensorMap{S,1,1} where {S<:ElementarySpace}
+
+"""
+Schmidt bond weight used in simple/cluster update
+"""
+struct SUWeight{E<:PEPSWeight}
+    x::Matrix{E}
+    y::Matrix{E}
+end
+
+function Base.size(wts::SUWeight)
+    @assert size(wts.x) == size(wts.y)
+    return size(wts.x)
+end
+
+function Base.eltype(wts::SUWeight)
+    @assert eltype(wts.x) == eltype(wts.y)
+    return eltype(wts.x)
+end
+
+function Base.:(==)(wts1::SUWeight, wts2::SUWeight)
+    return wts1.x == wts2.x && wts1.y == wts2.y
+end
+
+function Base.:(+)(wts1::SUWeight, wts2::SUWeight)
+    return SUWeight(wts1.x + wts2.x, wts1.y + wts2.y)
+end
+
+function Base.:(-)(wts1::SUWeight, wts2::SUWeight)
+    return SUWeight(wts1.x - wts2.x, wts1.y - wts2.y)
+end
+
+function Base.show(io::IO, wts::SUWeight)
+    N1, N2 = size(wts)
+    for (direction, r, c) in Iterators.product("xy", 1:N1, 1:N2)
+        println(io, "$direction[$r,$c]: ")
+        wt = (direction == 'x' ? wts.x[r, c] : wts.y[r, c])
+        for (k, b) in blocks(wt)
+            println(io, k, " = ", diag(b))
+        end
+    end
+end
+
+function Base.iterate(wts::SUWeight, state=1)
+    nx = prod(size(wts.x))
+    if 1 <= state <= nx
+        return wts.x[state], state + 1
+    elseif nx + 1 <= state <= 2 * nx
+        return wts.y[state - nx], state + 1
+    else
+        return nothing
+    end
+end
+
+function Base.length(wts::SUWeight)
+    @assert size(wts.x) == size(wts.y)
+    return 2 * prod(size(wts.x))
```

Comment:
This is slightly strange, in the sense that `length` here is the iteration length, while `size` is the size of the unit cell. Do you ever really use the iteration functionality? I would think it might be more convenient to have this behave as a "matrix" than as an iterator

Reply:
I really think that iteration is not the right abstraction for that. It sounds like these things really are mapping over all weights, while iteration is typically used to visit only a few, and to define an order. I don't really see that iterating first the `x` weights and then the `y` weights would constitute a natural order, so I would refrain from defining that.